src/parser/shaman/restoration/modules/talents/UnleashLife.js

```diff
@@ -55,128 +66,104 @@ class UnleashLife extends Analyzer {
-
-      if (hasUnleashLife) {
-        this.healing += calculateEffectiveHealing(event, UNLEASH_LIFE_HEALING_INCREASE);
-        this.unleashLifeHealRemaining - 1 < 0 ? this.unleashLifeHealRemaining = 0 : this.unleashLifeHealRemaining -= 1;
+    // These 3 heals only have 1 event and are handled easily
+    if (this.unleashLifeHealRemaining > 0 && ((spellId === SPELLS.HEALING_WAVE.id) || (spellId === SPELLS.HEALING_SURGE_RESTORATION.id) || (spellId === SPELLS.RIPTIDE.id && !event.tick))) {
+      this.healingBuff[spellId].healing += calculateEffectiveHealing(event, UNLEASH_LIFE_HEALING_INCREASE);
+      this.unleashLifeHealRemaining = 0;
+      debug && console.log("Heal:", spellId);
+
+      // I had to move the HoT application to the heal event as the buffapply event had too many false positives
+      if (spellId === SPELLS.RIPTIDE.id) {
+        this.healingBuff[spellId].playersActive.push(event.targetID);
+        debug && console.log("HoT Applied:", spellId, event.targetID);
       }
-    } else if (SPELLS.RIPTIDE.id === spellId && event.tick && this.buffedRiptideTarget === event.targetID) {
-      this.healing += calculateEffectiveHealing(event, UNLEASH_LIFE_HEALING_INCREASE);
-    } else if (spellId === SPELLS.UNLEASH_LIFE_TALENT.id) {
-      this.healing += event.amount + (event.absorbed || 0);
+
+      // Chain heal has up to 4 events, setting the variable to -1 to indicate that there might be more events coming
+    } else if (spellId === SPELLS.CHAIN_HEAL.id && (this.unleashLifeHealRemaining > 0 || (this.unleashLifeHealRemaining < 0 && this.buffedChainHealTimestamp + BUFFER_MS > event.timestamp))) {
+      this.healingBuff[spellId].healing += calculateEffectiveHealing(event, UNLEASH_LIFE_HEALING_INCREASE);
+      this.unleashLifeHealRemaining = -1;
+      this.buffedChainHealTimestamp = event.timestamp;
+      debug && console.log("Heal:", spellId);
     }
   }
 
+
   on_byPlayer_cast(event) {
     const spellId = event.ability.guid;
 
-    if(spellId === SPELLS.UNLEASH_LIFE_TALENT.id) {
+    if (spellId === SPELLS.UNLEASH_LIFE_TALENT.id) {
       this.unleashLifeCasts += 1;
-      this.unleashLifeRemaining += 1;
+      this.unleashLifeRemaining = true;
       this.lastUnleashLifeTimestamp = event.timestamp;
+      debug && console.log("New Unleash", event.timestamp);
     }
 
-    const hasUnleashLife = this.selectedCombatant.hasBuff(SPELLS.UNLEASH_LIFE_TALENT.id, event.timestamp, BUFFER_MS, BUFFER_MS);
+    if (this.unleashLifeRemaining && (this.lastUnleashLifeTimestamp + UNLEASH_LIFE_DURATION) <= event.timestamp) {
+      this.unleashLifeRemaining = false;
+      debug && console.log("Cast Timed out", event.timestamp);
+      return;
+    }
 
-    if(!hasUnleashLife) {
+    const hasUnleashLife = this.selectedCombatant.hasBuff(SPELLS.UNLEASH_LIFE_TALENT.id, event.timestamp, BUFFER_MS, BUFFER_MS);
+    if (!hasUnleashLife) {
```

Comment:
I'm confused about why another variable is needed.  You're tracking if the buff is up with unleashLifeRemaining, right?  What's requiring you to use a second check for if the buff is up?

Reply:
Must be leftover from the old solution. It essentially did nothing and I removed it.